File: src/main/java/leetcode/hot100/LeetCode141.java
package leetcode.hot100;

import common.ListNode;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/15 23:36
 */
public class LeetCode141 {

    public boolean hasCycle(ListNode head) {

        if (head == null || head.next == null) {
            return false;
        }
        ListNode slow = head;
        ListNode fast = head.next;

        //fix
        while (fast.next != null && fast.next.next != null) {

            if (slow == fast) {
                return true;
            }

            slow = slow.next;
            fast = fast.next.next;
        }

        return false;

    }
}

File: src/main/go/leetcode/lists/hot100/LeetCode141.go
package hot100

func hasCycle(head *ListNode) bool {
	if head == nil || head.Next == nil {
		return false
	}

	slow, fast := head, head.Next

	for fast != nil && fast.Next != nil {
		if slow == fast {
			return true
		}
		slow = slow.Next
		fast = fast.Next.Next
	}

	return false
}
